src/plugins/gs-plugin-external-apps.c

```diff
@@ -0,0 +1,804 @@
+		             GS_PLUGIN_ERROR_FAILED,
+		             "External asset's runtime member has no '%s' key "
+			     "set", JSON_RUNTIME_NAME_KEY);
+		return NULL;
+	}
+
+	json_url = json_object_get_string_member (runtime, JSON_RUNTIME_URL_KEY);
+	if (!json_url) {
+		g_set_error (error,
+		             GS_PLUGIN_ERROR,
+		             GS_PLUGIN_ERROR_FAILED,
+		             "External asset's runtime member has no '%s' key "
+			     "set", JSON_RUNTIME_URL_KEY);
+		return NULL;
+	}
+
+	*url = g_strdup (json_url);
+
+	return g_strdup (runtime_name);
+}
+
+static GsApp *
+gs_plugin_get_app_external_runtime (GsPlugin *plugin,
+				    GsApp *headless_app)
+{
+	GsApp *runtime = NULL;
+	GsPluginData *priv;
+	g_autofree char *id = NULL;
+	g_autofree char *full_id = NULL;
+	g_autofree char *url = NULL;
+	g_autofree char *json_data = NULL;
+	g_autoptr (GError) error = NULL;
+	const char *metadata;
+
+	metadata = gs_app_get_metadata_item (headless_app,
+					     METADATA_EXTERNAL_ASSETS);
+
+	if (!metadata)
+		return NULL;
+
+	json_data = g_uri_unescape_string (metadata, NULL);
+	id = extract_runtime_info_from_json_data (json_data, &url, &error);
+
+	if (!id) {
+		g_debug ("Error getting external runtime from "
+			 "metadata: %s", error->message);
+		return NULL;
+	}
+
+	priv = gs_plugin_get_data (plugin);
+
+	full_id = g_strdup_printf ("%s:%s",
+				   gs_flatpak_get_prefix (priv->flatpak),
+				   id);
+
+	runtime = gs_plugin_cache_lookup (plugin, full_id);
+	if (runtime) {
+		g_debug ("Found cached '%s'", full_id);
+		/* Adopt runtime if needed */
+		if (g_strcmp0 (gs_app_get_management_plugin (runtime),
+			       gs_plugin_get_name (plugin)) != 0) {
+			gs_app_set_management_plugin (runtime,
+						      gs_plugin_get_name (plugin));
+		}
+
+		return runtime;
+	}
+
+	runtime = gs_app_new (full_id);
+	gs_app_set_metadata (runtime, METADATA_HEADLESS_APP,
+			     gs_app_get_id (headless_app));
+	gs_app_set_metadata (runtime, METADATA_URL, url);
+	gs_app_set_metadata (runtime, "flatpak::kind", "runtime");
+	gs_app_set_kind (runtime, AS_APP_KIND_RUNTIME);
+	gs_app_set_flatpak_name (runtime, id);
+	gs_app_set_flatpak_arch (runtime, flatpak_get_default_arch ());
+	gs_app_set_flatpak_branch (runtime, "master");
+	gs_app_set_management_plugin (runtime, gs_plugin_get_name (plugin));
+
+	gs_plugin_cache_add (plugin, full_id, runtime);
+
+	if (gs_flatpak_is_installed (priv->flatpak, runtime, NULL, NULL)) {
+		gs_app_set_state (runtime, AS_APP_STATE_INSTALLED);
+	}
+
+	return runtime;
+}
+
+gboolean
+gs_plugin_refine_app (GsPlugin *plugin,
+		      GsApp *app,
+		      GsPluginRefineFlags flags,
+		      GCancellable *cancellable,
+		      GError **error)
+{
+	GsApp *ext_runtime;
+	const char *metadata = NULL;
+	GsPluginData *priv;
+
+	/* We cache all runtimes because an external runtime may have been
+	 * adopted by the flatpak plugins */
+	if (app_is_flatpak (app) &&
+	    g_strcmp0 (gs_app_get_flatpak_kind_as_str (app), "runtime") == 0) {
```

Comment:
The gs-flatpak header doesn't include the flatpak headers for returning the FLATPAK_REF_KIND and maybe that's intentional. So I prefer to add a *gs_flatpak_app_is_runtime* function returning a boolean instead (even if I use the flatpak.h header myself in gs-plugin-external-apps.c).